### DISCUSSION TITLE
Using a pre-compiled SNOPT package (trial version) with DAFoam.

### DISCUSSION THREAD

**Discussion by uttamcadambi07** [mdolab/dafoam#520] [General]
Hey everyone,

Just starting this thread so that those who need it can find it easily.

Requirements:
To use a pre-compiled SNOPT library with DAFoam as the alternate optimizer to IPOPT. The purpose for doing this is because Dr. He mentioned in this comment https://github.com/mdolab/dafoam/discussions/392#discussioncomment-7547569 that using IPOPT with ```fvSource``` as the design variable did not work. So this was the motivation.

Specs:
DAFoam v3.0.6
Ubuntu 18.04
OF v1812

Steps:

1. You need to download the pre-compiled files that are of the form ```libsnopt7.a``` and ```libsnopt7.so```. 
2. Place these in ```~/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/source```
3. You need to create a binding file. I have used the instructions on this link (note that it is for MAc but i will give the commands for linux) [Mac - https://github.com/mdolab/pyoptsparse/issues/335#issuecomment-1690100023] for linux, do this
```
python3 -m numpy.f2py -m snopt -c ~/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/source/f2py/snopt.pyf ~/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/*.f -L~/dafoam_v2/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/source -lsnopt7
```
4. This generates the binding file that looks like ```snopt.cpython-38-x86_64-linux-gnu.so```
5. create a directory called licenses and drop the *.lic file that the developers of SNOPT send to you in mail (usually valid for 3 months)
6. Set the environment variables: 
```
export PYOPTSPARSE_IMPORT_SNOPT_FROM=~/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/source/
export SNOPT_LICENSE=~/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/licenses/snopt7.lic
export LD_LIBRARY_PATH=~/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/source:$LD_LIBRARY_PATH
export PYTHONPATH=~/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/source:$PYTHONPATH

```

And then enjoy! It should work correctly now.

**Comment by friedenhe** [comment]
Thanks for sharing this!

**Comment by Ananth-Narayan-IITM** [comment]
I think this answer can be refined. I am facing issues. Upon following the instruction, I get the error when I run
```bash
python3 -m numpy.f2py -m snopt -c ~/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/source/f2py/snopt.pyf ~/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/*.f -L~/dafoam_v2/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/source -lsnopt7
```
the error as `error: [Errno 2] No such file or directory: '/home/dafoamuser/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/*.f'`

I requested the files from [UC San Diego](https://ccom.ucsd.edu/~optimizers/) and obtained these files
```
./libsnopt7.a
./libsnopt7.so
./snopt_cwrap.h
./snopt.h
./sqopt_cwrap.h
./sqopt.h
./snopt.lic
```

Further cloned the [v2.10.1](https://github.com/mdolab/pyoptsparse/releases/tag/v2.10.1) into `~/dafoam/repos/` such that I copied `libsnopt7.a` and `libsnopt7.so` to `/home/dafoamuser/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/source`

Then ran `python3 -m numpy.f2py -m snopt -c ~/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/source/f2py/snopt.pyf ~/dafoam/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/*.f -L~/dafoam_v2/repos/pyoptsparse-2.10.1/pyoptsparse/pySNOPT/source -lsnopt7` which gave me error as no file found in `/pySNOPT/*.f`

I suspect should we have `*.f` files under `/pySNOPT/`? Can you verify the steps and let me know in case of mistakes.

Thank you
